src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,877 @@
+        """
+
+        num_1d_tapers = tapers_1d.shape[-1]
+        tapers_1d = tapers_1d.astype(complex_type(self.dtype), copy=False)
+
+        blocks_mt = np.zeros(blocks[0, :, :].shape, dtype=self.dtype)
+
+        blocks_tapered = np.zeros(blocks[0, :, :].shape, dtype=complex_type(self.dtype))
+
+        taper_2d = np.zeros(
+            (blocks.shape[1], blocks.shape[2]), dtype=complex_type(self.dtype)
+        )
+
+        for ax1 in range(num_1d_tapers):
+            for ax2 in range(num_1d_tapers):
+                np.matmul(
+                    tapers_1d[:, ax1, np.newaxis],
+                    tapers_1d[:, ax2, np.newaxis].T,
+                    out=taper_2d,
+                )
+
+                for m in range(blocks.shape[0]):
+                    np.multiply(blocks[m, :, :], taper_2d, out=blocks_tapered)
+                    blocks_mt_post_fft = fft.fftn(blocks_tapered, axes=(-2, -1))
+                    blocks_mt += abs2(blocks_mt_post_fft)
+
+        blocks_mt /= blocks.shape[0] ** 2
+        blocks_mt /= tapers_1d.shape[0] ** 2
+
+        thon_rings = fft.fftshift(
+            blocks_mt
+        )  # max difference 10^-13, max relative difference 10^-14
+
+        return Image(thon_rings)
+
+    def elliptical_average(self, ffbbasis, thon_rings, circular):
+        """
+        Computes radial/elliptical average of the power spectrum
+
+        :param ffbbasis: FFBBasis instance.
+        :param thon_rings: Power spectrum.
+        :param circular: True for radial averaging and False for elliptical averaging.
+        :return: PSD and noise as 2-tuple of NumPy arrays.
+        """
+
+        # RCOPT, come back and change the indices for this method
+        coeffs_s = ffbbasis.evaluate_t(thon_rings).T
+        coeffs_n = coeffs_s.copy()
+
+        coeffs_s[np.argwhere(ffbbasis._indices["ells"] == 1)] = 0
+        if circular:
+            coeffs_s[np.argwhere(ffbbasis._indices["ells"] == 2)] = 0
+            noise = thon_rings
+        else:
+            coeffs_n[np.argwhere(ffbbasis._indices["ells"] == 0)] = 0
+            coeffs_n[np.argwhere(ffbbasis._indices["ells"] == 2)] = 0
+            noise = ffbbasis.evaluate(coeffs_n.T)
+
+        psd = ffbbasis.evaluate(coeffs_s.T)
+
+        return psd, noise
+
+    def background_subtract_1d(
+        self, thon_rings, linprog_method="interior-point", n_low_freq_cutoffs=14
+    ):
+        """
+        Estimate and subtract the background from the power spectrum
+
+        :param thon_rings: Estimated power spectrum
+        :param linprog_method: Method passed to linear progam solver (scipy.optimize.linprog).  Defaults to 'interior-point'.
+        :param n_low_freq_cutoffs: Low frequency cutoffs (loop iterations).
+        :return: 2-tuple of NumPy arrays (PSD after noise subtraction and estimated noise)
+        """
+
+        # compute radial average
+        center = thon_rings.shape[-1] // 2
+
+        if thon_rings.ndim == 3:
```

Comment:
The code path currently used is fed from the result of `ffbbasis.evaluate`, which should be a stack of 1 (if not it is an error..).  So this is only being called with `== 3` case right now .
 
I think it is reasonable to produce a descriptive error here if someone is tinkering around/extending and passes in something that does not make sense. (A stack greater than one or wrong dim).  I say that because a lot of this logic might silently work against those cases and either crash or make junk someplace farther away from the error... I would hope we save someone time.

I can do the zero slice in the prior function and document shape in the docstring if that is more appropriate. LMK, trying to hit all these today... thanks.

Reply:
Ok that makes sense. Let's leave this then.